flink/src/main/java/com/zhihu/flink/source/tidb/TiDBInputFormat.java

```diff
@@ -0,0 +1,211 @@
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.typeutils.ResultTypeQueryable;
+import org.apache.flink.api.java.typeutils.RowTypeInfo;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.core.io.GenericInputSplit;
+import org.apache.flink.core.io.InputSplit;
+import org.apache.flink.core.io.InputSplitAssigner;
+import org.apache.flink.table.api.DataTypes;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.table.types.utils.TypeConversions;
+import org.apache.flink.types.Row;
+
+public class TiDBInputFormat extends RichInputFormat<Row, InputSplit> implements ResultTypeQueryable<Row> {
+
+  private String pdAddresses;
+
+  private String databaseName;
+
+  private String tableName;
+
+  private String[] fieldNames;
+
+  private DataType[] fieldTypes;
+
+  private List<SplitInternal> splits;
+
+  private List<ColumnHandleInternal> columnHandleInternals;
+
+  private transient RecordCursorInternal cursor;
+
+  private transient ClientSession clientSession;
+
+
+  /**
+   * see {@link TiDBInputFormat#builder()}
+   */
+  private TiDBInputFormat() {
+  }
+
+  @Override
+  public void configure(Configuration parameters) {
+    // do nothing here
+  }
+
+  @Override
+  public BaseStatistics getStatistics(BaseStatistics cachedStatistics) throws IOException {
+    return cachedStatistics;
+  }
+
+  @Override
+  public InputSplit[] createInputSplits(int minNumSplits) throws IOException {
+    GenericInputSplit[] inputSplits = new GenericInputSplit[splits.size()];
+    for (int i = 0; i < inputSplits.length; i++) {
+      inputSplits[i] = new GenericInputSplit(i, inputSplits.length);
+    }
+    return inputSplits;
+  }
+
+  @Override
+  public InputSplitAssigner getInputSplitAssigner(InputSplit[] inputSplits) {
+    return new DefaultInputSplitAssigner(inputSplits);
+  }
+
+  @Override
+  public void openInputFormat() throws IOException {
+    clientSession = new ClientSession(new ClientConfig(pdAddresses));
+  }
+
+  @Override
+  public void closeInputFormat() throws IOException {
+    // TODO clientSession will auto close?
+  }
+
+  @Override
+  public void open(InputSplit split) throws IOException {
+    SplitInternal splitInternal = splits.get(split.getSplitNumber());
+    RecordSetInternal recordSetInternal = new RecordSetInternal(clientSession, splitInternal,
+        columnHandleInternals, Optional.empty());
+    cursor = recordSetInternal.cursor();
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (cursor != null) {
+      cursor.close();
+    }
+  }
+
+  @Override
+  public boolean reachedEnd() throws IOException {
+    return !cursor.advanceNextPosition();
+  }
+
+  @Override
+  public Row nextRecord(Row row) throws IOException {
+    for (int i = 0; i < row.getArity(); i++) {
+      DataType fieldType = fieldTypes[i];
+      Object object;
+      // map datatype
+      if (fieldType.equals(DataTypes.FLOAT())) {
```

Comment:
Prepared accessors would be better than many if else statements which is error prone and fragile.